File: 505-The Maze II.py
from heapq import heappush, heappop
class Solution(object):
    def shortestDistance(self, maze, start, destination):
        """
        :type maze: List[List[int]]
        :type start: List[int]
        :type destination: List[int]
        :rtype: int
        """
# Dijkstra 1
        visited = set()
        row, col = len(maze), len(maze[0])
        pq = []
        heappush(pq, (0, start[0], start[1]))
        
        while pq:
            step, currR, currC = heappop(pq)
            if currR == destination[0] and currC == destination[1]:
                    return step
            visited.add((currR, currC))
            for dr, dc in ((1, 0), (0, 1), (-1, 0), (0, -1)):
                R, C, currStep = currR, currC, step
                while 0<=R+dr<row and 0<=C+dc<col and maze[R+dr][C+dc] == 0:
                    R += dr
                    C += dc
                    currStep += 1
                if (R, C) not in visited:
                    heappush(pq, (currStep, R, C))
        return -1
    
    
# Dijkstra 2
        row, col = len(maze), len(maze[0])
        pq = []
        dist = [[float('inf')]*col for _ in range(row)]
        heappush(pq, (0, start[0], start[1]))
        dist[start[0]][start[1]] = 0
        
        while pq:
            step, currR, currC = heappop(pq)
            for dr, dc in ((1, 0), (0, 1), (-1, 0), (0, -1)):
                R, C, count = currR, currC, 0
                while 0<=R+dr<row and 0<=C+dc<col and maze[R+dr][C+dc] == 0:
                    R += dr
                    C += dc
                    count += 1
                if dist[currR][currC] + count < dist[R][C]:
                    dist[R][C] = dist[currR][currC] + count
                    heappush(pq, (dist[currR][currC] + count, R, C)) 
        if dist[destination[0]][destination[1]] == float('inf'):
            return -1
        else:
            return dist[destination[0]][destination[1]]

File: C++/505-The Maze II.cpp
// first AC
class Solution {
public:
    int d[5] = {1, 0, -1, 0, 1};
    int shortestDistance(vector<vector<int>>& maze, vector<int>& start, vector<int>& destination) {
        int row = maze.size(), col = maze[0].size();
        vector<vector<int>> visited(row, vector<int>(col, INT_MAX));

        queue<tuple<int, int, int>> q;
        q.push({start[0], start[1], 0});
        visited[start[0]][start[1]] = 0;
        
        while (!q.empty()) {
            auto [currR, currC, dist] = q.front();
            q.pop();

            if (currR == destination[0] && currC == destination[1]) {
                visited[currR][currC] = min(visited[currR][currC], dist);
                continue;
            }

            for (int i = 0; i < 4; i++) {
                int nxtR = currR, nxtC = currC;
                while (nxtR >= 0 && nxtR < row && nxtC >= 0 && nxtC < col && maze[nxtR][nxtC] == 0) {
                    nxtR += d[i];
                    nxtC += d[i + 1];
                }
                nxtR -= d[i];
                nxtC -= d[i + 1];
                
                int moved = abs(nxtR-currR) + abs(nxtC-currC);

                if (visited[nxtR][nxtC] > dist + moved) {
                    visited[nxtR][nxtC] = dist + moved;
                    q.push({nxtR, nxtC, dist + moved});
                }
            }
        }


        return visited[destination[0]][destination[1]] == INT_MAX?-1:visited[destination[0]][destination[1]];
    }
};


// Better, with a small modification
class Solution {
public:
    int d[5] = {1, 0, -1, 0, 1};
    int shortestDistance(vector<vector<int>>& maze, vector<int>& start, vector<int>& destination) {
        int row = maze.size(), col = maze[0].size();
        vector<vector<int>> visited(row, vector<int>(col, INT_MAX));

        queue<tuple<int, int, int>> q;
        q.push({start[0], start[1], 0});
        
        while (!q.empty()) {
            auto [currR, currC, dist] = q.front();
            q.pop();

            if (dist > visited[currR][currC]) continue;   
            // Not dist >= visited[currR][currC] here because we have visited[nxtR][nxtC] = dist + moved below

            if (currR == destination[0] && currC == destination[1]) {
                visited[currR][currC] = min(visited[currR][currC], dist);
                continue;
            }

            for (int i = 0; i < 4; i++) {
                int nxtR = currR, nxtC = currC;
                while (nxtR >= 0 && nxtR < row && nxtC >= 0 && nxtC < col && maze[nxtR][nxtC] == 0) {
                    nxtR += d[i];
                    nxtC += d[i + 1];
                }
                nxtR -= d[i];
                nxtC -= d[i + 1];
                
                int moved = abs(nxtR-currR) + abs(nxtC-currC);

                if (visited[nxtR][nxtC] > dist + moved) {
                    visited[nxtR][nxtC] = dist + moved;
                    q.push({nxtR, nxtC, dist + moved});
                }
            }
        }


        return visited[destination[0]][destination[1]] == INT_MAX?-1:visited[destination[0]][destination[1]];
    }
};